subprojects/core/src/main/java/org/gradle/api/internal/DefaultDomainObjectCollection.java

```diff
@@ -154,7 +186,15 @@ public void whenObjectRemoved(Closure action) {
      * Adds an action which is executed before this collection is mutated. Any exception thrown by the action will veto the mutation.
      */
     public void beforeChange(Runnable action) {
-        mutateActions.add(action);
+        if (mutateAction == null) {
```

Comment:
Let's make this an `Action` and move this logic to `Actions` so we can reuse it elsewhere.